scripts/collect_activations.py

```diff
@@ -1,77 +1,169 @@
+
+        # Build input tokens
+        input_ids = torch.cat(
+            (
+                tokenized_fsp,
+                tokenized_question,
+                tokenized_instruction,
+                tokenized_response,
+            ),
+            dim=0,
+        )
 
-    model = AutoModelForCausalLM.from_pretrained(model_name).cuda()
+        locs_to_cache = []
+        for loc_type in loc_types_to_cache:
+            if loc_type == "resp":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp)
+                        + len(tokenized_question)
+                        + len(tokenized_instruction),
+                        len(input_ids),
+                    )
+                )
+            elif loc_type == "instr":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp) + len(tokenized_question),
+                        len(tokenized_fsp)
+                        + len(tokenized_question)
+                        + len(tokenized_instruction),
+                    )
+                )
+            elif loc_type == "quest":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp), len(tokenized_fsp) + len(tokenized_question)
+                    )
+                )
+            elif loc_type == "fsp":
+                locs_to_cache.extend(range(len(tokenized_fsp)))
 
-    layers_to_cache = list(range(model.config.num_hidden_layers))
-    activations_by_question: list[QuestionActivations] = []
-    for question in tqdm(questions):
-        locs_to_cache = set()
-        for key, locs in question.locs.items():
-            locs_to_cache.update(locs)
-        locs_to_cache = sorted(locs_to_cache)
-
-        # # hack to add "Let's think step by step"
-        # min_loc = locs_to_cache[0]
-        # # hack to not add anything else
-        # locs_to_cache = list(range(min_loc - 10, min_loc))  # + locs_to_cache
-        # # hack to have smaller prompt and faster inference
-        # input_ids = torch.tensor(question.tokens[:min_loc])
-        input_ids = torch.tensor(question.tokens)
         resid_acts = clean_run_with_cache(
             model, input_ids, layers_to_cache, locs_to_cache
         )
 
-        activations_by_question.append(QuestionActivations(resid_acts, locs_to_cache))
+        for layer_idx, layer in enumerate(layers_to_cache):
+            activations_by_layer[layer].append(resid_acts[layer_idx].cpu())
 
-    model_name = model_name.replace("/", "--")
-    resid_acts_path = (
-        f"results/activations_{model_name}_{task_name}_S{seed}_N{num_samples}.pkl"
-    )
+    # Create dir for activations
+    acts_type = "+".join(list(loc_types_to_cache))
+    acts_name = f"acts_{acts_type}"
+    acts_folder_path = os.path.join(output_folder, acts_name)
+    os.makedirs(acts_folder_path, exist_ok=True)
```

Comment:
I don't like that, I would prefer to have one folder for acts on question + instructions and one folder for acts on response